File: Chapter06/MoveComponent.cpp
//
//  MoveComponent.cpp
//  Game-mac
//
//  Created by Michael Lehmann on 12.01.21.
//  Copyright © 2021 Michael Lehmann. All rights reserved.
//

#include "MoveComponent.h"
#include "Actor.h"
#include "Math.h"

MoveComponent::MoveComponent(class Actor* owner, int updateOrder)
    :Component(owner, updateOrder)
{
}

void MoveComponent::Update(float deltaTime)
{
    if (!Math::NearZero(mAngularSpeed))
    {
        Quaternion rot = mOwner->GetRotation();
        float angle = mAngularSpeed * deltaTime;
        
        // Create quaternion for incremental rotation
        // (Rotate about up axis)
        Quaternion inc(Vector3::UnitZ, angle);
        
        // Concatenate old and new quaternion
        rot = Quaternion::Concatenate(rot, inc);
    
        mOwner->SetRotation(rot);
    }
    
    // Updating position based on forward speed stays the same
    if (!Math::NearZero(mForwardSpeed))
    {
        Vector3 pos = mOwner->GetPosition();
        pos += mOwner->GetForward() * mForwardSpeed * deltaTime;
        
        mOwner->SetPosition(pos);
    }
}

File: Chapter06_Veldrid/MoveComponent.cs
using System;
using System.Numerics;

namespace Chapter06
{
    public class MoveComponent : Component
    {
        public MoveComponent(Actor owner, int updateOrder = 10)
            : base(owner, updateOrder)
        {
        }

        public float AngularSpeed { get; set; }

        public float ForwardSpeed { get; set; }

        public override void Update(float deltaTime)
        {
            if (!AngularSpeed.NearZero())
            {
                Quaternion rotation = Owner.Rotation;
                float angle = AngularSpeed * deltaTime;

                // Create quaternion for incremental rotation
                // (Rotate about up axis)
                Quaternion inc = new Quaternion(Vector3.UnitZ, angle);
                rotation = Quaternion.Concatenate(rotation, inc);

                // Concatenate old and new quaternion
                Owner.Rotation = rotation;
            }

            // Updating position based on forward speed stays the same
            if (!ForwardSpeed.NearZero())
            {
                Vector3 position = Owner.Position;
                position += Owner.Forward * ForwardSpeed * deltaTime;

                Owner.Position = position;
            }
        }
    }
}